File: mycelix-workspace/sdk-ts/src/storage/router.ts
/**
 * UESS Classification Router
 *
 * Maps E/N/M classification to storage decisions.
 * @see docs/architecture/uess/UESS-01-CORE.md §4
 */

import {
  type StorageTier,
  type StorageBackend,
  MutabilityMode,
  AccessControlMode,
} from './types.js';
import { EmpiricalLevel, NormativeLevel, MaterialityLevel } from '../epistemic/types.js';

import type { EpistemicClassification } from '../epistemic/types.js';

// =============================================================================
// Router Configuration
// =============================================================================

/**
 * Configuration for the classification router
 */
export interface RouterConfig {
  /** Default TTL for M0 in ms (default: 1 hour) */
  m0DefaultTtl: number;

  /** Default TTL for M1 in ms (default: 7 days) */
  m1DefaultTtl: number;

  /** Default DHT replication factor */
  dhtDefaultReplication: number;

  /** Minimum replication for M3 */
  m3MinReplication: number;

  /** Maximum replication for M3 */
  m3MaxReplication: number;

  /** Enable IPFS for N2+ content-addressed data */
  enableIpfs: boolean;

  /** Enable Filecoin for M3 archival */
  enableFilecoin: boolean;
}

/**
 * Default router configuration
 */
export const DEFAULT_ROUTER_CONFIG: RouterConfig = {
  m0DefaultTtl: 60 * 60 * 1000, // 1 hour
  m1DefaultTtl: 7 * 24 * 60 * 60 * 1000, // 7 days
  dhtDefaultReplication: 10,
  m3MinReplication: 5,
  m3MaxReplication: 30,
  enableIpfs: true,
  enableFilecoin: false, // Requires Filecoin integration
};

// =============================================================================
// Classification Router Interface
// =============================================================================

/**
 * Classification Router - Maps E/N/M to storage decisions
 */
export interface ClassificationRouter {
  /**
   * Determine storage tier for classification
   */
  route(classification: EpistemicClassification): StorageTier;

  /**
   * Check if reclassification requires migration
   */
  requiresMigration(
    from: EpistemicClassification,
    to: EpistemicClassification
  ): boolean;

  /**
   * Get all backends that should hold this data
   */
  getBackends(classification: EpistemicClassification): StorageBackend[];

  /**
   * Calculate replication factor
   */
  getReplication(
    classification: EpistemicClassification,
    networkSize: number
  ): number;

  /**
   * Validate classification transition (must be monotonically increasing)
   */
  validateTransition(
    from: EpistemicClassification,
    to: EpistemicClassification
  ): { valid: boolean; error?: string };
}

// =============================================================================
// Storage Router Implementation
// =============================================================================

/**
 * Default implementation of the classification router
 */
export class StorageRouter implements ClassificationRouter {
  constructor(
    private readonly config: RouterConfig = DEFAULT_ROUTER_CONFIG,
    private readonly networkSizeProvider: () => number = () => 100
  ) {}

  /**
   * Route classification to storage tier
   */
  route(classification: EpistemicClassification): StorageTier {
    const { empirical, normative, materiality } = classification;

    // Determine backend from M-level
    const backend = this.getBackendForMateriality(materiality, normative);
    const additionalBackends = this.getAdditionalBackends(materiality, normative, empirical);

    // Determine mutability from E-level
    const mutability = this.getMutabilityForEmpirical(empirical);

    // Determine access control from N-level
    const accessControl = this.getAccessControlForNormative(normative);

    // Calculate replication
    const networkSize = this.networkSizeProvider();
    const replication = this.getReplication(classification, networkSize);

    // Encryption required for N0/N1 (INV-5)
    const encrypted = normative < NormativeLevel.N2_Network;

    // Content-addressed for E3+ (CID-based)
    const contentAddressed = empirical >= EmpiricalLevel.E3_Cryptographic;

    // TTL for M0/M1
    const ttl = this.getTTL(materiality);

    return {
      backend,
      additionalBackends,
      replication,
      mutability,
      accessControl,
      ttl,
      encrypted,
      contentAddressed,
    };
  }

  /**
   * Get primary backend based on M-level
   */
  private getBackendForMateriality(
    materiality: MaterialityLevel,
    _normative: NormativeLevel
  ): StorageBackend {
    switch (materiality) {
      case MaterialityLevel.M0_Ephemeral:
        return 'memory';

      case MaterialityLevel.M1_Temporal:
        return 'local';

      case MaterialityLevel.M2_Persistent:
        // DHT is primary for all M2 data
        return 'dht';

      case MaterialityLevel.M3_Immutable:
        // Filecoin > IPFS > DHT for immutable data
        // IPFS is content-addressed and immutable by nature, ideal for M3
        if (this.config.enableFilecoin) return 'filecoin';
        if (this.config.enableIpfs) return 'ipfs';
        return 'dht';

      default:
        return 'local';
    }
  }

  /**
   * Get additional backends for redundancy
   */
  private getAdditionalBackends(
    materiality: MaterialityLevel,
    normative: NormativeLevel,
    empirical: EmpiricalLevel
  ): StorageBackend[] {
    const backends: StorageBackend[] = [];

    if (materiality === MaterialityLevel.M3_Immutable) {
      // M3 uses DHT index + IPFS for retrieval
      backends.push('dht');
      if (this.config.enableIpfs && normative >= NormativeLevel.N2_Network) {
        backends.push('ipfs');
      }
    } else if (materiality === MaterialityLevel.M2_Persistent) {
      // M2 with N2+ can add IPFS for content-addressing (E3+)
      if (
        normative >= NormativeLevel.N2_Network &&
        empirical >= EmpiricalLevel.E3_Cryptographic &&
        this.config.enableIpfs
      ) {
        backends.push('ipfs');
      }
      // Local backup for N0/N1
      if (normative < NormativeLevel.N2_Network) {
        backends.push('local');
      }
    }

    return backends;
  }

  /**
   * Calculate replication factor
   */
  getReplication(
    classification: EpistemicClassification,
    networkSize: number
  ): number {
    const { materiality, normative } = classification;

    switch (materiality) {
      case MaterialityLevel.M0_Ephemeral:
        return 0; // No replication

      case MaterialityLevel.M1_Temporal:
        return 1; // Local only

      case MaterialityLevel.M2_Persistent:
        switch (normative) {
          case NormativeLevel.N0_Personal:
            return 3; // Personal data, moderate replication
          case NormativeLevel.N1_Communal:
            return 5; // Community data, higher replication
          default:
            return this.config.dhtDefaultReplication;
        }

      case MaterialityLevel.M3_Immutable:
        // Logarithmic replication for M3
        return this.calculateLogarithmicReplication(networkSize);

      default:
        return 1;
    }
  }

  /**
   * Logarithmic Replication for M3
   *
   * Formula: Replication = ceil(log₂(NetworkSize))
   *
   * Ensures:
   * - Small networks: Adequate redundancy (log₂(100) = 7)
   * - Large networks: No bloat (log₂(1M) = 20)
   */
  private calculateLogarithmicReplication(networkSize: number): number {
    const { m3MinReplication, m3MaxReplication } = this.config;

    if (networkSize <= 0) return m3MinReplication;

    const logReplication = Math.ceil(Math.log2(networkSize));

    return Math.max(m3MinReplication, Math.min(m3MaxReplication, logReplication));
  }

  /**
   * Get mutability mode from E-level
   */
  private getMutabilityForEmpirical(empirical: EmpiricalLevel): MutabilityMode {
    if (empirical <= EmpiricalLevel.E1_Testimonial) {
      return MutabilityMode.MUTABLE_CRDT;
    }
    if (empirical === EmpiricalLevel.E2_PrivateVerify) {
      return MutabilityMode.APPEND_ONLY;
    }
    return MutabilityMode.IMMUTABLE;
  }

  /**
   * Get access control mode from N-level
   */
  private getAccessControlForNormative(normative: NormativeLevel): AccessControlMode {
    switch (normative) {
      case NormativeLevel.N0_Personal:
        return AccessControlMode.OWNER;
      case NormativeLevel.N1_Communal:
        return AccessControlMode.CAPBAC;
      default:
        return AccessControlMode.PUBLIC;
    }
  }

  /**
   * Get TTL for M-level
   */
  private getTTL(materiality: MaterialityLevel): number | undefined {
    switch (materiality) {
      case MaterialityLevel.M0_Ephemeral:
        return this.config.m0DefaultTtl;
      case MaterialityLevel.M1_Temporal:
        return this.config.m1DefaultTtl;
      default:
        return undefined; // No expiry
    }
  }

  /**
   * Check if reclassification requires migration
   */
  requiresMigration(
    from: EpistemicClassification,
    to: EpistemicClassification
  ): boolean {
    const fromTier = this.route(from);
    const toTier = this.route(to);

    // Migration required if backend changes
    if (fromTier.backend !== toTier.backend) return true;

    // Or if encryption status changes
    if (fromTier.encrypted !== toTier.encrypted) return true;

    // Or if mutability mode changes (more restrictive)
    if (fromTier.mutability !== toTier.mutability) return true;

    // Or if content-addressing changes
    if (fromTier.contentAddressed !== toTier.contentAddressed) return true;

    return false;
  }

  /**
   * Get all backends for classification
   */
  getBackends(classification: EpistemicClassification): StorageBackend[] {
    const tier = this.route(classification);
    return [tier.backend, ...tier.additionalBackends];
  }

  /**
   * Validate classification transition (INV-2: monotonically increasing)
   */
  validateTransition(
    from: EpistemicClassification,
    to: EpistemicClassification
  ): { valid: boolean; error?: string } {
    // E-level must be >= (cannot downgrade verification)
    if (to.empirical < from.empirical) {
      return {
        valid: false,
        error: `Cannot downgrade E-level from E${from.empirical} to E${to.empirical}`,
      };
    }

    // N-level must be >= (cannot narrow consensus scope)
    if (to.normative < from.normative) {
      return {
        valid: false,
        error: `Cannot downgrade N-level from N${from.normative} to N${to.normative}`,
      };
    }

    // M-level must be >= (cannot reduce durability)
    if (to.materiality < from.materiality) {
      return {
        valid: false,
        error: `Cannot downgrade M-level from M${from.materiality} to M${to.materiality}`,
      };
    }

    return { valid: true };
  }
}

/**
 * Global router instance with default configuration
 */
export const defaultRouter = new StorageRouter();

/**
 * Create a router with custom configuration
 */
export function createRouter(
  config?: Partial<RouterConfig>,
  networkSizeProvider?: () => number
): StorageRouter {
  return new StorageRouter(
    { ...DEFAULT_ROUTER_CONFIG, ...config },
    networkSizeProvider
  );
}

File: mycelix-workspace/sdk/src/storage/router.rs
// Copyright (C) 2024-2026 Tristan Stoltz / Luminous Dynamics
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial licensing: see COMMERCIAL_LICENSE.md at repository root
//! UESS Storage Router
//!
//! Routes data to appropriate storage backends based on E/N/M classification.
//! Implements INV-2 (monotonic classification transitions).

use super::types::{AccessControlMode, MutabilityMode, StorageBackend, StorageTier};
use crate::epistemic::{EmpiricalLevel, EpistemicClassification, MaterialityLevel, NormativeLevel};
use serde::{Deserialize, Serialize};

/// Router configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RouterConfig {
    /// TTL for M0 (ephemeral) data in milliseconds
    pub m0_ttl_ms: u64,
    /// TTL for M1 (temporal) data in milliseconds
    pub m1_ttl_ms: u64,
    /// Default replication factor for DHT
    pub dht_default_replication: usize,
    /// Minimum replication for M3 (immutable)
    pub m3_min_replication: usize,
    /// Maximum replication for M3
    pub m3_max_replication: usize,
    /// Whether IPFS backend is enabled
    pub enable_ipfs: bool,
    /// Whether Filecoin backend is enabled
    pub enable_filecoin: bool,
    /// Network size for replication calculation
    pub network_size: usize,
}

impl Default for RouterConfig {
    fn default() -> Self {
        Self {
            m0_ttl_ms: 3600 * 1000,          // 1 hour
            m1_ttl_ms: 7 * 24 * 3600 * 1000, // 7 days
            dht_default_replication: 10,
            m3_min_replication: 5,
            m3_max_replication: 30,
            enable_ipfs: true,
            enable_filecoin: false,
            network_size: 100,
        }
    }
}

/// Storage router for E/N/M classification → tier mapping
#[derive(Debug, Clone)]
pub struct StorageRouter {
    config: RouterConfig,
}

impl StorageRouter {
    /// Create a new router with configuration
    pub fn new(config: RouterConfig) -> Self {
        Self { config }
    }

    /// Create with default configuration
    pub fn default_router() -> Self {
        Self::new(RouterConfig::default())
    }

    /// Route classification to storage tier
    ///
    /// This is the core routing logic that maps E/N/M levels to storage behavior.
    pub fn route(&self, classification: &EpistemicClassification) -> StorageTier {
        // 1. Backend from Materiality
        let (backend, additional_backends) =
            self.backend_from_materiality(classification.materiality);

        // 2. Mutability from Empirical
        let mutability = self.mutability_from_empirical(classification.empirical);

        // 3. Access control from Normative
        let access_control = self.access_control_from_normative(classification.normative);

        // 4. Encryption required for N0/N1
        let encrypted = matches!(
            classification.normative,
            NormativeLevel::N0Personal | NormativeLevel::N1Communal
        );

        // 5. Content addressing for E3+
        let content_addressed = matches!(
            classification.empirical,
            EmpiricalLevel::E3Cryptographic | EmpiricalLevel::E4PublicRepro
        );

        // 6. TTL from Materiality
        let ttl_ms = self.ttl_from_materiality(classification.materiality);

        // 7. Replication factor
        let replication = self.replication_factor(classification);

        StorageTier {
            backend,
            additional_backends,
            replication,
            mutability,
            access_control,
            ttl_ms,
            encrypted,
            content_addressed,
        }
    }

    /// Determine backend from Materiality level
    fn backend_from_materiality(
        &self,
        materiality: MaterialityLevel,
    ) -> (StorageBackend, Vec<StorageBackend>) {
        match materiality {
            MaterialityLevel::M0Ephemeral => (StorageBackend::Memory, vec![]),
            MaterialityLevel::M1Temporal => (StorageBackend::Local, vec![]),
            MaterialityLevel::M2Persistent => (StorageBackend::DHT, vec![StorageBackend::Local]),
            MaterialityLevel::M3Foundational => {
                if self.config.enable_ipfs {
                    if self.config.enable_filecoin {
                        (
                            StorageBackend::IPFS,
                            vec![StorageBackend::Filecoin, StorageBackend::DHT],
                        )
                    } else {
                        (StorageBackend::IPFS, vec![StorageBackend::DHT])
                    }
                } else {
                    (StorageBackend::DHT, vec![StorageBackend::Local])
                }
            }
        }
    }

    /// Determine mutability mode from Empirical level
    fn mutability_from_empirical(&self, empirical: EmpiricalLevel) -> MutabilityMode {
        match empirical {
            EmpiricalLevel::E0Null | EmpiricalLevel::E1Testimonial => MutabilityMode::MutableCRDT,
            EmpiricalLevel::E2PrivateVerify => MutabilityMode::AppendOnly,
            EmpiricalLevel::E3Cryptographic | EmpiricalLevel::E4PublicRepro => {
                MutabilityMode::Immutable
            }
        }
    }

    /// Determine access control from Normative level
    fn access_control_from_normative(&self, normative: NormativeLevel) -> AccessControlMode {
        match normative {
            NormativeLevel::N0Personal => AccessControlMode::Owner,
            NormativeLevel::N1Communal => AccessControlMode::CapBAC,
            NormativeLevel::N2Network | NormativeLevel::N3Axiomatic => AccessControlMode::Public,
        }
    }

    /// Determine TTL from Materiality level
    fn ttl_from_materiality(&self, materiality: MaterialityLevel) -> Option<u64> {
        match materiality {
            MaterialityLevel::M0Ephemeral => Some(self.config.m0_ttl_ms),
            MaterialityLevel::M1Temporal => Some(self.config.m1_ttl_ms),
            MaterialityLevel::M2Persistent | MaterialityLevel::M3Foundational => None,
        }
    }

    /// Calculate replication factor
    fn replication_factor(&self, classification: &EpistemicClassification) -> usize {
        match classification.materiality {
            MaterialityLevel::M0Ephemeral => 0, // No replication for ephemeral
            MaterialityLevel::M1Temporal => 1,  // Local only
            MaterialityLevel::M2Persistent => {
                match classification.normative {
                    NormativeLevel::N0Personal => 3, // Moderate for personal
                    NormativeLevel::N1Communal => 5, // Higher for communal
                    _ => self.config.dht_default_replication,
                }
            }
            MaterialityLevel::M3Foundational => {
                // Logarithmic replication based on network size
                // Formula: ceil(log2(network_size)), bounded by [min, max]
                let log_replication = (self.config.network_size as f64).log2().ceil() as usize;
                log_replication.clamp(
                    self.config.m3_min_replication,
                    self.config.m3_max_replication,
                )
            }
        }
    }

    /// Validate a classification transition (INV-2: monotonic only)
    ///
    /// Returns Ok(()) if transition is valid, Err with reason if not.
    pub fn validate_transition(
        &self,
        from: &EpistemicClassification,
        to: &EpistemicClassification,
    ) -> Result<(), TransitionError> {
        // Empirical level must not decrease
        if (to.empirical as u8) < (from.empirical as u8) {
            return Err(TransitionError::EmpiricalDowngrade {
                from: from.empirical,
                to: to.empirical,
            });
        }

        // Normative level must not decrease
        if (to.normative as u8) < (from.normative as u8) {
            return Err(TransitionError::NormativeDowngrade {
                from: from.normative,
                to: to.normative,
            });
        }

        // Materiality level must not decrease
        if (to.materiality as u8) < (from.materiality as u8) {
            return Err(TransitionError::MaterialityDowngrade {
                from: from.materiality,
                to: to.materiality,
            });
        }

        Ok(())
    }

    /// Check if a transition requires migration (backend/encryption change)
    pub fn requires_migration(
        &self,
        from: &EpistemicClassification,
        to: &EpistemicClassification,
    ) -> bool {
        let from_tier = self.route(from);
        let to_tier = self.route(to);

        // Migration required if:
        // 1. Backend changes
        // 2. Encryption status changes
        // 3. Mutability mode changes
        // 4. Content addressing changes
        from_tier.backend != to_tier.backend
            || from_tier.encrypted != to_tier.encrypted
            || from_tier.mutability != to_tier.mutability
            || from_tier.content_addressed != to_tier.content_addressed
    }

    /// Get configuration
    pub fn config(&self) -> &RouterConfig {
        &self.config
    }

    /// Update network size (for replication calculation)
    pub fn set_network_size(&mut self, size: usize) {
        self.config.network_size = size;
    }
}

/// Classification transition error
#[derive(Debug, Clone, thiserror::Error)]
pub enum TransitionError {
    /// Attempted to downgrade empirical level (violates monotonicity)
    #[error("Cannot downgrade Empirical level from {from:?} to {to:?}")]
    EmpiricalDowngrade {
        /// Original empirical level
        from: EmpiricalLevel,
        /// Attempted downgraded level
        to: EmpiricalLevel,
    },
    /// Attempted to downgrade normative level (violates monotonicity)
    #[error("Cannot downgrade Normative level from {from:?} to {to:?}")]
    NormativeDowngrade {
        /// Original normative level
        from: NormativeLevel,
        /// Attempted downgraded level
        to: NormativeLevel,
    },
    /// Attempted to downgrade materiality level (violates monotonicity)
    #[error("Cannot downgrade Materiality level from {from:?} to {to:?}")]
    MaterialityDowngrade {
        /// Original materiality level
        from: MaterialityLevel,
        /// Attempted downgraded level
        to: MaterialityLevel,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    fn classification(
        e: EmpiricalLevel,
        n: NormativeLevel,
        m: MaterialityLevel,
    ) -> EpistemicClassification {
        EpistemicClassification {
            empirical: e,
            normative: n,
            materiality: m,
        }
    }

    #[test]
    fn test_backend_from_materiality() {
        let router = StorageRouter::default_router();

        // M0 → Memory
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.backend, StorageBackend::Memory);

        // M1 → Local
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        ));
        assert_eq!(tier.backend, StorageBackend::Local);

        // M2 → DHT
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M2Persistent,
        ));
        assert_eq!(tier.backend, StorageBackend::DHT);

        // M3 → IPFS (if enabled)
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M3Foundational,
        ));
        assert_eq!(tier.backend, StorageBackend::IPFS);
    }

    #[test]
    fn test_mutability_from_empirical() {
        let router = StorageRouter::default_router();

        // E0-E1 → MutableCRDT
        let tier = router.route(&classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.mutability, MutabilityMode::MutableCRDT);

        // E2 → AppendOnly
        let tier = router.route(&classification(
            EmpiricalLevel::E2PrivateVerify,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.mutability, MutabilityMode::AppendOnly);

        // E3-E4 → Immutable
        let tier = router.route(&classification(
            EmpiricalLevel::E3Cryptographic,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.mutability, MutabilityMode::Immutable);
    }

    #[test]
    fn test_access_control_from_normative() {
        let router = StorageRouter::default_router();

        // N0 → Owner
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.access_control, AccessControlMode::Owner);

        // N1 → CapBAC
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N1Communal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.access_control, AccessControlMode::CapBAC);

        // N2-N3 → Public
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N2Network,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.access_control, AccessControlMode::Public);
    }

    #[test]
    fn test_encryption_from_normative() {
        let router = StorageRouter::default_router();

        // N0 → encrypted
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert!(tier.encrypted);

        // N1 → encrypted
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N1Communal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert!(tier.encrypted);

        // N2+ → not encrypted
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N2Network,
            MaterialityLevel::M0Ephemeral,
        ));
        assert!(!tier.encrypted);
    }

    #[test]
    fn test_content_addressing_from_empirical() {
        let router = StorageRouter::default_router();

        // E0-E2 → not content addressed
        let tier = router.route(&classification(
            EmpiricalLevel::E2PrivateVerify,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert!(!tier.content_addressed);

        // E3-E4 → content addressed
        let tier = router.route(&classification(
            EmpiricalLevel::E3Cryptographic,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert!(tier.content_addressed);
    }

    #[test]
    fn test_ttl_from_materiality() {
        let router = StorageRouter::default_router();

        // M0 → has TTL
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert!(tier.ttl_ms.is_some());

        // M2-M3 → no TTL
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M2Persistent,
        ));
        assert!(tier.ttl_ms.is_none());
    }

    #[test]
    fn test_replication_factor() {
        let router = StorageRouter::default_router();

        // M0 → 0 (no replication)
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M0Ephemeral,
        ));
        assert_eq!(tier.replication, 0);

        // M2/N0 → moderate (3)
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M2Persistent,
        ));
        assert_eq!(tier.replication, 3);

        // M2/N1 → higher (5)
        let tier = router.route(&classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N1Communal,
            MaterialityLevel::M2Persistent,
        ));
        assert_eq!(tier.replication, 5);
    }

    #[test]
    fn test_transition_validation() {
        let router = StorageRouter::default_router();

        // Valid upgrade
        let from = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        let to = classification(
            EmpiricalLevel::E2PrivateVerify,
            NormativeLevel::N1Communal,
            MaterialityLevel::M2Persistent,
        );
        assert!(router.validate_transition(&from, &to).is_ok());

        // Invalid: E downgrade
        let to_bad_e = classification(
            EmpiricalLevel::E0Null,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        assert!(router.validate_transition(&from, &to_bad_e).is_err());

        // Invalid: N downgrade
        let from_n1 = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N1Communal,
            MaterialityLevel::M1Temporal,
        );
        let to_n0 = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        assert!(router.validate_transition(&from_n1, &to_n0).is_err());

        // Invalid: M downgrade
        let from_m2 = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M2Persistent,
        );
        let to_m1 = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        assert!(router.validate_transition(&from_m2, &to_m1).is_err());
    }

    #[test]
    fn test_requires_migration() {
        let router = StorageRouter::default_router();

        // Same classification → no migration
        let c = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        assert!(!router.requires_migration(&c, &c));

        // M1 → M2: backend change → migration
        let from = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        let to = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M2Persistent,
        );
        assert!(router.requires_migration(&from, &to));

        // N0 → N2: encryption change → migration
        let from = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        let to = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N2Network,
            MaterialityLevel::M1Temporal,
        );
        assert!(router.requires_migration(&from, &to));

        // E1 → E3: mutability change → migration
        let from = classification(
            EmpiricalLevel::E1Testimonial,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        let to = classification(
            EmpiricalLevel::E3Cryptographic,
            NormativeLevel::N0Personal,
            MaterialityLevel::M1Temporal,
        );
        assert!(router.requires_migration(&from, &to));
    }
}
